pandas/core/groupby/groupby.py

```diff
@@ -880,16 +883,57 @@ def _python_apply_general(
             Function to apply
         data : Series or DataFrame
             Data to apply f to
+        is_transform : bool, default False
+            Indicator for whether the function is actually a transform
+            and should not have group keys prepended. This is used
+            in _make_wrapper which generates both transforms (e.g. diff)
+            and non-transforms (e.g. corr)
+        is_empty_agg : bool, default False
+            Indicator for whether the function is actually an aggregation
+            on an empty result. We don't want to warn for this case.
+            See _GroupBy._python_agg_general.
 
         Returns
         -------
         Series or DataFrame
             data after applying f
         """
         keys, values, mutated = self.grouper.apply(f, data, self.axis)
+        not_indexed_same = mutated or self.mutated
+        override_group_keys = False
+
+        if (not not_indexed_same and self.group_keys is lib.no_default) and not (
```

Comment:
can you comment here on when this is hit (in words)

Reply:
When all of the following are true:
 - the resulting index is the same as the input index;
 - `group_keys` is left unspecified in the call to `.groupby(...)`;
 - the function is not in transformation_kernels (e.g. `cumsum` or `shift`); and
 - the result is either not computed by `_python_agg_general` or is nonempty.

I am wondering if instead of the last bullet, we should just not warn/override on any empty result. Will investigate.